docs/inventory_plugins.md

```diff
@@ -0,0 +1,197 @@
+Inventory updates change from using scripts which are vendored as executable
+python scripts in the AWX folder `awx/plugins/inventory` (taken originally from
+Ansible folder `contrib/inventory`) to using dynamically-generated
+YAML files which conform to the specifications of the `auto` inventory plugin
+which are then parsed by their respective inventory plugin.
+
+The major organizational change is that the inventory plugins are
+part of the Ansible core distribution, whereas the same logic used to
+be a part of AWX source.
+
+## Prior Background for Transition
+
+AWX used to maintain logic that parsed `.ini` inventory file contents,
+in addition to interpreting the JSON output of scripts, re-calling with
+the `--host` option in the case the `_meta.hostvars` key was not provided.
+
+### Switch to Ansible Inventory
+
+The CLI entry point `ansible-inventory` was introduced in Ansible 2.4.
+In Tower 3.2, inventory imports began running this command
+as an intermediary between the inventory and
+the import's logic to save content to database. Using `ansible-inventory`
+eliminates the need to maintain source-specific logic,
+relying on Ansible's code instead. This also allows us to
+count on a consistent data structure outputted from `ansible-inventory`.
+There are many valid structures that a script can provide, but the output
+from `ansible-inventory` will always be the same,
+thus the AWX logic to parse the content is simplified.
+This is why even scripts must be ran through the `ansible-inventory` CLI.
+
+Along with this switchover, a backported version of
+`ansible-inventory` was provided that supported Ansible versions 2.2 and 2.3.
+
+### Removal of Backport
+
+In AWX 3.0.0 (and Tower 3.5), the backport of `ansible-inventory`
+was removed, and support for using custom virtual environments was added.
+This set the minimum version of Ansible necessary to run _any_
+inventory update to 2.4.
+
+## Inventory Plugin Versioning
+
+Beginning in Ansible 2.5, cloud sources in Ansible started migrating
+away from "contrib" scripts (meaning they lived in the contrib folder)
+to the inventory plugin model.
+
+In AWX 4.0.0 (and Tower 3.5) inventory source types start to switchover
+to plugins, provided that sufficient compatibility is in place for
+the version of Ansible present in the custom virtualenv where the import
+is being ran.
+
+To see what version the plugin transition will happen, see
+`awx/main/models/inventory.py` and look for the source name as a
+subclass of `PluginFileInjector`, and there should be an `initial_version`
+which is the first version that testing deemed to have sufficient parity
+in the content its inventory plugin returns. For example, `openstack` will
+begin using the inventory plugin in Ansible version 2.8, because the
+openstack inventory plugin had an issue with sending logs to stdout which
+was fixed in that version. If you run an openstack inventory update in
+2.8 or lower, it will use the script.
+
```

Comment:
This could be rewritten some for greater clarity.

Reply:
yeah, I'm working on some touchups following up with Ryan's suggestions. That paragraph:

> To see what version the plugin transition will happen, see
`awx/main/models/inventory.py` and look for the source name as a
subclass of `PluginFileInjector`, and there should be an `initial_version`
which is the first version that testing deemed to have sufficient parity
in the content its inventory plugin returns. For example, `openstack` will
begin using the inventory plugin in Ansible version 2.8.
If you run an openstack inventory update with Ansible
2.7.x or lower, it will use the script.